armv7a9-zynq7000/zynq.c

```diff
@@ -119,6 +116,271 @@ struct {
+                                                             ((!!clk->pll.srcsel) << 4);
+			break;
+
+		case ctrl_gem0_clk:
+		case ctrl_gem1_clk:
+			id = clk->dev - ctrl_gem0_clk;
+			*(zynq_common.slcr + slcr_gem0_clk_ctrl + id) = (*(zynq_common.slcr + slcr_gem0_clk_ctrl + id) & ~0x03f03f71) | (!!clk->pll.clkact0) |
+                                                            ((clk->pll.srcsel & 0x7) << 4) | ((clk->pll.divisor0 & 0x3f) << 8) | ((clk->pll.divisor1 & 0x3f) << 20);
+			break;
+
+		case ctrl_smc_clk:
+			*(zynq_common.slcr + slcr_smc_clk_ctrl) = (*(zynq_common.slcr + slcr_smc_clk_ctrl) & ~0x00003f31) | (!!clk->pll.clkact0) |
+                                                      ((clk->pll.srcsel & 0x3) << 4) | ((clk->pll.divisor0 & 0x3f) << 8);
+			break;
+
+		case ctrl_lqspi_clk:
+			*(zynq_common.slcr + slcr_lqspi_clk_ctrl) = (*(zynq_common.slcr + slcr_lqspi_clk_ctrl) & ~0x00003f31) | (!!clk->pll.clkact0) |
+                                                        ((clk->pll.srcsel & 0x3) << 4) | ((clk->pll.divisor0 & 0x3f) << 8);
+			break;
+
+		case ctrl_sdio_clk:
+			*(zynq_common.slcr + slcr_sdio_clk_ctrl) = (*(zynq_common.slcr + slcr_sdio_clk_ctrl) & ~0x00003f33) | (!!clk->pll.clkact0) | ((!!clk->pll.clkact1) << 1) |
+                                                       ((clk->pll.srcsel & 0x3) << 4) | ((clk->pll.divisor0 & 0x3f) << 8);
+			break;
+
+		case ctrl_uart_clk:
+			*(zynq_common.slcr + slcr_uart_clk_ctrl) = (*(zynq_common.slcr + slcr_uart_clk_ctrl) & ~0x00003f33) | (!!clk->pll.clkact0) |
+                                                       ((!!clk->pll.clkact1) << 1) | ((clk->pll.srcsel & 0x3) << 4) | ((clk->pll.divisor0 & 0x3f) << 8);
+			break;
+
+		case ctrl_spi_clk:
+			*(zynq_common.slcr + slcr_spi_clk_ctrl) = (*(zynq_common.slcr + slcr_spi_clk_ctrl) & ~0x00003f33) | (!!clk->pll.clkact0) |
+													  ((!!clk->pll.clkact1) << 1) | ((clk->pll.srcsel & 0x3) << 4) | ((clk->pll.divisor0 & 0x3f) << 8);
+			break;
+
+		case ctrl_can_clk:
+			*(zynq_common.slcr + slcr_can_clk_ctrl) = (*(zynq_common.slcr + slcr_can_clk_ctrl) & ~0x03f03f33) | (!!clk->pll.clkact0) |
+                                                      ((!!clk->pll.clkact1) << 1) | ((clk->pll.srcsel & 0x3) << 4) | ((clk->pll.divisor0 & 0x3f) << 8) |
+                                                      ((clk->pll.divisor1 & 0x3f) << 20);
+			break;
+
+		case ctrl_can_mioclk:
+			*(zynq_common.slcr + slcr_can_mioclk_ctrl) = (*(zynq_common.slcr + slcr_can_mioclk_ctrl) & ~0x007f007f) | (clk->mio.mux0 & 0x3f) | ((!!clk->mio.ref0) << 6) |
+                                                         ((clk->mio.mux1 & 0x3f) << 16) | ((!!clk->mio.ref1) << 22);
+			break;
+
+		default:
+			_zynq_slcrLock();
+			return ERR_ARG;
+	}
+
+	_zynq_slcrLock();
+
+	return 0;
+}
+
+
+int _zynq_getCtlClock(ctl_clock_t *clk)
+{
+	u32 id;
+	u32 val = 0;
+
+	_zynq_slcrUnlock();
```

Comment:
Do we need to unlock when reading a value?

Reply:
Good note, I've tested it and it works without unlocking slcr.